File: web/src/api/mixer.ts
// Channel-baseband spectrum stream client. Mirrors WS /api/v1/diag/mixer
// (see internal/api/mixer.go). Same connect/reconnect scaffolding the
// symbol stream uses; the wire shape is one JSON MixerFrame per message.

import { type ClientConfig } from "./client";
import {
  openReconnectingSocket,
  type SocketStatus,
} from "./reconnectingSocket";

export interface MixerFrame {
  ts_ns: number;
  center_hz: number;
  sample_rate_hz: number;
  offset_hz: number;
  // The receiver's residual carrier-offset estimate (Hz). The tuned view
  // is the raw window re-mixed by this, so a locked loop centres it.
  carrier_offset_hz: number;
  // FFT-shifted dBFS bins of the channelized baseband. raw_bins shows the
  // carrier at its residual offset; tuned_bins shows it after the
  // carrier-recovery correction. Both equal length; index 0 is the lowest
  // frequency, index n/2 is DC.
  raw_bins: number[];
  tuned_bins: number[];
}

export type MixerFrameHandler = (f: MixerFrame) => void;
export type StatusHandler = (s: SocketStatus) => void;

export interface MixerStream {
  close(): void;
}

export interface MixerOptions {
  serial: string;
  // Receiver selector: "p25-c4fm" or "p25-cqpsk". Default "p25-c4fm".
  proto?: string;
  // Frequency offset in Hz, relative to the SDR centre, tuned down to
  // baseband before channelizing. Default 0.
  offset?: number;
  /** Fired once when the stream gives up on this device; re-enumerate. */
  onGone?: () => void;
  onFrame: MixerFrameHandler;
  onStatus?: StatusHandler;
}

export function mixerWebSocketURL(cfg: ClientConfig, opts: MixerOptions): string {
  const params = new URLSearchParams({ device: opts.serial });
  if (opts.proto) params.set("proto", opts.proto);
  if (opts.offset != null && opts.offset !== 0)
    params.set("offset", String(Math.round(opts.offset)));
  const u = new URL(
    `/api/v1/diag/mixer?${params.toString()}`,
    cfg.baseURL || window.location.href,
  );
  u.protocol = u.protocol === "https:" ? "wss:" : "ws:";
  return u.toString();
}

export function openMixerStream(
  cfg: ClientConfig,
  opts: MixerOptions,
): MixerStream {
  return openReconnectingSocket({
    url: () => mixerWebSocketURL(cfg, opts),
    onStatus: opts.onStatus,
    onGone: opts.onGone,
    onMessage: (data) => {
      try {
        const frame = JSON.parse(data) as MixerFrame;
        if (frame && Array.isArray(frame.tuned_bins)) {
          opts.onFrame(frame);
        }
      } catch {
        // Drop malformed.
      }
    },
  });
}

File: internal/api/mixer.go
package api

import (
	"context"
	"encoding/json"
	"net/http"
	"time"

	"github.com/gorilla/websocket"
)

// MixerFrame is the wire shape of one channel-baseband spectrum snapshot
// for the web Mixer plot (OP25's Raw Mixer / Tuned Mixer tabs). It
// mirrors symbolscope.MixerFrame so the api package stays free of a
// dependency on the DSP package; the MixerProvider interface bridges the
// two.
//
// RawBins is the FFT of the channelized baseband as the receiver sees it
// (the carrier sits at the residual offset); TunedBins is the same window
// re-mixed by the receiver's carrier-offset estimate so a locked loop
// centres the carrier. Both are dBFS, FFT-shifted (Bins[0] is the lowest
// frequency, Bins[n/2] is DC), equal length.
type MixerFrame struct {
	TimestampNs     int64     `json:"ts_ns"`
	CenterHz        uint32    `json:"center_hz"`
	SampleRateHz    uint32    `json:"sample_rate_hz"`
	OffsetHz        int32     `json:"offset_hz"`
	CarrierOffsetHz float64   `json:"carrier_offset_hz"`
	RawBins         []float32 `json:"raw_bins"`
	TunedBins       []float32 `json:"tuned_bins"`
}

// MixerProvider is the daemon-side abstraction the mixer endpoint
// consumes. The daemon implements it on top of the iqtap broker map +
// internal/scanner/symbolscope; tests substitute a fake.
type MixerProvider interface {
	// OpenMixerStream starts a per-request symbol-recovery engine on the
	// named device with the Mixer FFT taps enabled. proto selects the
	// receiver ("p25-c4fm" / "p25-cqpsk"); offsetHz tunes an off-centre
	// channel down to baseband before channelizing. Returns the wire-frame
	// channel and a cleanup func the caller MUST invoke on disconnect.
	OpenMixerStream(ctx context.Context, serial, proto string, offsetHz int32) (<-chan MixerFrame, func(), error)
}

// handleMixerStream answers WS /api/v1/diag/mixer?device=...&proto=...&offset=....
//
// Frames arrive as JSON text messages. Server pings every 30 s to keep
// proxies from idling the socket. Connection stays open until the client
// disconnects or the device disappears. Mirrors handleSymbolStream.
func (s *Server) handleMixerStream(w http.ResponseWriter, r *http.Request) {
	if s.mixer == nil {
		s.writeError(w, http.StatusServiceUnavailable, "mixer subsystem not enabled")
		return
	}
	q := r.URL.Query()
	serial := q.Get("device")
	if serial == "" {
		s.writeError(w, http.StatusBadRequest, "device query parameter is required")
		return
	}
	proto := q.Get("proto")
	if proto == "" {
		proto = "p25-c4fm"
	}
	if !symbolProtos[proto] {
		s.writeError(w, http.StatusBadRequest, "unsupported proto (want p25-c4fm or p25-cqpsk)")
		return
	}
	offset := int32(parseIntQuery(q, "offset", 0, -30_000_000, 30_000_000))

	upgrader := wsUpgrader
	if s.cors.enabled() {
		cors := s.cors
		upgrader.CheckOrigin = func(req *http.Request) bool {
			origin := req.Header.Get("Origin")
			if origin == "" {
				return true
			}
			_, ok := cors.originAllowed(origin)
			return ok
		}
	}
	// Resolve the device BEFORE the upgrade. An unknown serial must be an HTTP
	// status, not a successful handshake followed by a close: browser clients
	// reset their reconnect backoff on `onopen`, so a handshake that always
	// succeeds makes the backoff dead code and a stale tab reconnects forever.
	if s.rejectUnknownDevice(w, serial) {
		return
	}
	conn, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		s.log.Debug("api: mixer WS upgrade failed", "err", err)
		return
	}
	defer conn.Close()

	ctx, cancel := context.WithCancel(r.Context())
	defer cancel()

	frames, cleanup, err := s.mixer.OpenMixerStream(ctx, serial, proto, offset)
	if err != nil {
		s.log.Debug("api: mixer OpenMixerStream failed", "serial", serial, "err", err)
		_ = conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(
			websocket.CloseInternalServerErr, err.Error()))
		return
	}
	defer cleanup()

	// Drain client → server messages so close frames are processed.
	go func() {
		for {
			if _, _, err := conn.NextReader(); err != nil {
				cancel()
				return
			}
		}
	}()

	ping := time.NewTicker(30 * time.Second)
	defer ping.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-ping.C:
			_ = conn.SetWriteDeadline(time.Now().Add(5 * time.Second))
			if err := conn.WriteMessage(websocket.PingMessage, nil); err != nil {
				return
			}
		case f, ok := <-frames:
			if !ok {
				return
			}
			body, err := json.Marshal(f)
			if err != nil {
				s.log.Debug("api: mixer frame marshal", "err", err)
				continue
			}
			_ = conn.SetWriteDeadline(time.Now().Add(5 * time.Second))
			if err := conn.WriteMessage(websocket.TextMessage, body); err != nil {
				return
			}
		}
	}
}
